src/PostgREST/App.hs

```diff
@@ -248,18 +246,26 @@ app dbStructure conf apiRequest =
                   singular = contentType == CTSingularJSON
                   paramsAsSingleObject = iPreferSingleObjectParameter apiRequest
               row <- H.query () $
-                callProc qi prms returnsScalar q cq topLevelRange shouldCount
+                callProc qi prms returnsScalar q cq shouldCount
                          singular paramsAsSingleObject
                          (contentType == CTTextCSV)
                          (contentType == CTOctetStream) _isReadOnly bField
-              let (tableTotal, queryTotal, body) =
-                    fromMaybe (Just 0, 0, "[]") row
+                         (pgVersion dbStructure)
+              let (tableTotal, queryTotal, body, jsonHeaders) =
+                    fromMaybe (Just 0, 0, "[]", "[]") row
                   (status, contentRange) = rangeHeader queryTotal tableTotal
-              if singular && queryTotal /= 1
-                then do
-                  HT.condemn
-                  return $ singularityError (toInteger queryTotal)
-                else return $ responseLBS status [toHeader contentType, contentRange] (toS body)
+                  decodedHeaders = first toS $ eitherDecode $ toS jsonHeaders :: Either Text [GucHeader]
+              case decodedHeaders of
+                Left _ -> return gucHeadersError
+                Right hs -> do
+                  let customHeaders = toHeaders hs
+                      contentTypeH | isNothing $ lookup hContentType customHeaders = [toHeader contentType]
```

Comment:
Too bad duplicate keys are significant or we could make customHeaders a map like

```haskell
let customHeaders = M.fromList (toHeaders hs)
      `M.union` M.fromList [toHeader contentType, contentRange]
```

And then use `toList` in `responseLBS`. ...but yeah duplicate keys are allowed.

___

In any case I've got a question about the current logic. It falls back to the server<->client negotiated content type if the GUC has not set a content type, otherwise uses the GUC's choice. However `responseContentTypeOrError` recognizes only a short list of Accept types for rpc calls:

```haskell
ActionInvoke _ ->  [CTApplicationJSON, CTSingularJSON, CTTextCSV, CTOctetStream]
```

If a client calls `/rpc/generate_page` with `Accept: text/html` won't the request die when `responseContentTypeOrError` returns Left -- even though `generate_page` may be capable of producing HTML?

It seems to be only when clients add `*/*` to their Accept header that RPC calls will work with functions returning fancy content types.

Reply:
I was thinking of this situation:

```haskell
it "content negotiation allows nonstandard Accept when function agrees" $
  request methodGet "/rpc/hello_in_text_plain" (acceptHdrs "text/plain") ""
    `shouldRespondWith`
    "hello"
    { matchHeaders = ["Content-Type" <:> "text/plain"] }
```
Which produces
```
status mismatch:
  expected: 200
  but got:  415
missing header:
  Content-Type: text/plain
the actual headers were:
  Content-Type: application/json; charset=utf-8
body mismatch:
  expected: hello
  but got:  {"message":"None of these Content-Types are available: text/plain"}
```

My thought is, the client should be able to request the content type that the RPC will ultimately provide.

Reply:
@begriffs you are thinking about this PR from the direction of "let's generate html with postgrest" which is not the goal for it, i know i mentioned in other threads that this type of PR might open the html feature (which i think is a bad way of doing it) but it does not mean that this particular pr should make it possible.

There are lots of other useful headers out there (like set cookie which will enable cookie auth) that have nothing to do with the incoming request headers, and this PR is immensely useful already without having to implement Accept/Content-Type correlation.

Let's merge this and leave the html thing for another time (that is if someone wants to implement it).

Reply:
HTML is one example, but really I was thinking it makes sense that a client can request a certain content type and not be denied if the stored proc does indeed return that type. Having to request `application/octet-stream` and then getting some other type back is nonstandard HTTP. Requesting `Accept: application/octet-stream,*/*` and getting something else back does make sense.

I can see the point that this may not end up being a common case, and it may be hard to fix this aspect of the code because usually the code evaluates the Accept header against a fixed set of types and decides immediately to fail rather than running the stored proc and checking retroactively. We could overlook this in the current pull request, but it does mean our treatment of HTTP is kind of broken.

Reply:
i would not say the current way is broken, it's only that this PR allows a RPC to break it. If the RPC does not touch content-type then everything works as it should, so if it breaks, it's on "them" :)

Reply:
come to think of it, isn't it the same in every other env php/node, the browser sends an accept header while the script can do whatever it wants?

the only thing different here is that we don't allow "unknown" types to fall through to the rpc but this is a completely different topic, how do we know if the rpc is going to take care of the content type or not (probably by inspecting the return type).

or maybe an RPC should not be allowed to directly change the content type header and the content type should be determined by it's return type. like bytea -> octetstream, text -> (json or html), any other -> as it was before.
but as i said, this is a completely different discussion/pr.

Reply:
Initially I thought that the responsibility of making the http treatment adequate should be left to the developer and didn't consider the unsupported case where client does an `Accept: text/html` and the response has the same content type returned, was always thinking in terms of an `Accept: application/octet-stream` and then RPC overwrites content-type but as mentioned that is not an adequate treatment of http.

I think I have a solution for this and preserve the most of http semantics.

What if we allow arbitrary `Content-Types` to be requested on RPC? Well the RPC could set the matching content type but the body could be anything, including the default json, we cannot check the body for compliance but really the only way for the body to be say an `text/html` is that the return type of the RPC function is bytea/text, both of these types being supported by the binary output feature, so we could enforce that condition.

So the solution I propose would be:

- Make RPC support other arbitrary content types, with the requirement that when one of these types is requested, the function has to return `bytea/text` or client selects a column of that type for a row, same condition used for binary output.
- Check that the `Accept` content type matches the one provided by the RPC custom content type otherwise fail.

However as @ruslantalpa mentioned the feature request was mostly about setting other headers and not `Content-Type`, @begriffs how about if I remove the content type overwriting from this PR and then continue this discussion on another issue? I could open one with my proposed solution when this is merged.

Reply:
I'll go ahead with removing the `Content-Type` overwriting(I'll keep a local branch with the change and I'll make a note for opening another issue with the proposal).